code/game/objects/items/weapons/flamethrower.dm

```diff
@@ -15,239 +126,267 @@
+		to_chat(user, "<span class='userdanger'>You need both hands free to attack with \the [src]!</span>")
+		return
+	if(lit && status == 2) //we want it to have different attack values depending on what status and if it's on or not
+		damtype = BURN
+		force = force_welder
+		armour_penetration = armour_penetration_welder
+		hitsound = 'sound/items/welder.ogg'
+		consume_fuel(consumption_rate)
+	else
+		damtype = BRUTE
+		force = initial(force)
+		armour_penetration = initial(armour_penetration)
+		hitsound = "swing_hit"
+	..()
+
+/obj/item/flamethrower/advanced/afterattack(atom/target, mob/user, flag)
+	. = ..()
+	if(flag)
+		return // too close
+	if(user.get_inactive_hand())
+		to_chat(user, "<span class='userdanger'>You need both hands free to use \the [src]!</span>")
+		return
+	if(lit)
+		if(status == 2)
+			return
+		else if(user && user.get_active_hand() == src) // Make sure our user is still holding us
+			playsound(get_turf(src),'sound/magic/fireball.ogg', 200, TRUE)
+			fireline(target, user)
+			consume_fuel(consumption_rate)
+
+obj/item/flamethrower/advanced/tool_use_check(mob/living/user)
+	if(!lit)
+		to_chat(user, "<span class='notice'>[src] has to be on to complete this task!</span>")
+		return FALSE
+	return TRUE
+
+obj/item/flamethrower/advanced/use_tool(atom/target, mob/living/user, delay, amount, volume, datum/callback/extra_checks)
+	. = ..()
+	consume_fuel(consumption_rate)
+
+/obj/item/flamethrower/advanced/AltClick(mob/user)
+	if(lit)
+		to_chat(user, "<span class='notice'> turn off [src] first!</span>")
+	else if(ptank && isliving(user) && user.Adjacent(src))
+		user.put_in_hands(ptank)
+		ptank = null
+		to_chat(user, "<span class='notice'>You remove the plasma tank from [src]!</span>")
+		update_icon()
+
+obj/item/flamethrower/advanced/CtrlClick(mob/user)
+	playsound(get_turf(user), 'sound/items/change_jaws.ogg', 50, 1)
+	if(status == 1)
+		status = 2
+		consumption_rate = consumption_rate_welder
+		tool_behaviour = TOOL_WELDER
+		to_chat(user, "<span class='notice'> You swap to welder mode.</span>")
+	else
+		status = 1
+		consumption_rate = initial(consumption_rate)
+		tool_behaviour = null
+		to_chat(user, "<span class='notice'> You swap to flamethrower mode.</span>")
+	update_icon()
+
+/obj/item/flamethrower/advanced/examine(mob/user)
+	. = ..()
+	if(ptank)
+		. += "<span class='notice'>[src] has \a [ptank] attached. Alt-click to remove it.</span>"
+	if(status == 2)
+		. += "<span class='notice'>[src] is in welder mode. Ctrl-click to swap modes.</span>"
+	else
+		. += "<span class='notice'>[src] is in flamethrower mode. Ctrl-click to swap modes.</span>"
+
+/obj/item/flamethrower/advanced/Initialize(mapload)
+	. = ..()
+	if(create_with_tank)
+		ptank = new /obj/item/tank/plasma/full(src)
+	update_icon()
+
+/obj/item/flamethrower/advanced/hit_reaction(mob/living/carbon/human/owner, atom/movable/hitby, attack_text = "the attack", final_block_chance = 0, damage = 0, attack_type = MELEE_ATTACK)
```

Comment:
This proc mostly uses the same logic as the basic one. Combine them where possible